crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -961,10 +974,24 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
-    fn module_ty_from_name(&self, module_name: Option<ModuleName>) -> Type<'db> {
-        module_name
-            .and_then(|module_name| resolve_module(self.db, module_name))
-            .map_or(Type::Unknown, |module| Type::Module(module.file()))
+    fn module_ty_from_name(
+        &mut self,
+        module_name: Option<ModuleName>,
+        node: AnyNodeRef,
+    ) -> Type<'db> {
+        let Some(module_name) = module_name else {
+            return Type::Unknown;
+        };
+
+        if let Some(module) = resolve_module(self.db, module_name.clone()) {
+            Type::Module(module.file())
+        } else {
+            if let Some(builder) = self.push_diagnostic(node, "reportMissingImport") {
+                builder.finish(format!("Import '{module_name}' could not be resolved."));
+            }
```

Comment:
Fair enough. It just feels slightly more verbose because it now requires an extra `push_diagnostic` call, and this needs to be repeated at every call site. 

The overall idea is that the API is similar to `map.entry` where we want to defer the insertion and also already pre-set the required values (the key).

Reply:
I simplified the API to use `format_args` to defer the expensive message allocation for now. We may need to go back to a builder-like pattern in the future. But that depends on how we design our diagnostics